### ISSUE TITLE
Calling now after map blows up

### ISSUE THREAD

**Issue by sake92** [rescala-lang/REScala#33]
Hello! Thank you for this awesome library! :)

When I do this in ScalaJS (JVM is the same..):
```scala
object Main {

  def main(args: Array[String]): Unit = {

    val bla$ = Signal { "whatever" }

    val nesto = bla$.map { e =>
      content()
    }
    println("DONE", nesto.now)
  }

  val isEdit$ = Var(false)
  def content() = {
    val e2 = isEdit$.map(x => x)
    println(e2.now) // HERE!
  }
}
```

I get:

```
Throwables.scala:18 Uncaught java.lang.IllegalStateException: ba.sake.rxtags.example.Main.main nesto:17(rescala.levelbased.LevelState@4) has an error value
    at $f_Lrescala_reactives_Signal__readValueOnce__Lrescala_core_Scheduler__O (file:///C:/projects/sake/RxTags/examples/target/scala-2.13/examples-fastopt.js:9332:63)
    at $c_Lba_sake_rxtags_example_Main$.main__AT__V (file:///C:/projects/sake/RxTags/examples/target/scala-2.13/examples-fastopt.js:780:29)
    at $s_Lba_sake_rxtags_example_Main__main__AT__V (file:///C:/projects/sake/RxTags/examples/target/scala-2.13/examples-fastopt.js:721:38)
    at file:///C:/projects/sake/RxTags/examples/target/scala-2.13/examples-fastopt.js:24188:1
    at file:///C:/projects/sake/RxTags/examples/target/scala-2.13/examples-fastopt.js:24189:4
```

Any idea how to fix it?

**Comment by rmgk** [comment]
The core issue you have is that you should not use `.now` inside transactions – that is inside of any signal or event that is updating.

If you look at the Java Stacktrace (which are still kind of hard to read, but better than JS) you get the following:
Exception in thread "main" rescala.operator.RExceptions$ObservedException: »tests.rescala.misc.Main.main nesto:12(rescala.scheduler.levelbased.LevelState@536aaa8d)« was accessd but contained an exception: 
java.util.NoSuchElementException: 
tests.rescala.misc.Main.content e2:20(rescala.scheduler.levelbased.LevelState@17579e0f) is empty

This basically says that `nesto` contained an exception (which you see because you call `println("DONE", nesto.now)`) and `nesto` says that the exception is because `e2` is empty when it is accessed at `println(e2.now)`

The reason `e2` is empty is because when you create the `nesto` signal, it is initialized inside of a transaction by executing the content function. The content function then creates a new signal `e2` inside the same transaction, you immediately read the value of that signal which is not defined until the end of the transaction, thus “empty”.

There are several ways to fix this the easiest but hackiest is to just inline the content function:
```
    val nesto = bla$.map { e =>
      val e2 = isEdit$.map(x => x)
      println(e2.now) // HERE!
    }
```
This kinda works, but there is no relationship between `e2` and nesto, because `.now` does not create a dependency.
It also only works because some macro magic inside of REScala rewrites your code to fix the initialization of the signal inside the nested call.
In general, you should not create and use a signal inside the same outer signal.

To propose a clean solution the full code then is:

```
import rescala.default._

object Main {

  def main(args: Array[String]): Unit = {

    val bla$ = Signal { "whatever" }

    val isEdit$ = Var(false)
    val e2      = isEdit$.map(x => x)

    val nesto = bla$.map { e =>
      e2.value
    }

    nesto.observe(println)

    println(s"DONE ${nesto.now}")
  }

}
```
Note that this uses `e2.value` inside of the `map` call to access the value of the signal. It also uses `nesto.observe(println)` to automatically print the value of `nesto` whenever it changes (and when the observer is first added)

If you really want to create signals inside signals, do not use them inside the signal that created them, consider this example:

```
import rescala.default._

object Main {

  def main(args: Array[String]): Unit = {

    val bla$ = Signal { "whatever" }

    val isEdit$ = Var(false)

    val nesto = bla$.map { e =>
      val e2 = isEdit$.map(x => x)
      e2
    }

    val accessed = Signal.dynamic {
      nesto.value.value
    }

    val accessed2 = nesto.flatten

    accessed.observe(println)
    accessed2.observe(println)

    println(s"DONE ${accessed.now}")
    println(s"DONE ${accessed2.now}")
  }

}
```

basically, `Signal.dynamic` allows you to use dynamic signals here accessed using `.value.value` to unwrap twice, alternatively you can also use `flatten` which does the same thing. 

Hope this helps.
